api/filrewardsd/service/service.go

```diff
@@ -0,0 +1,658 @@
+package service
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"math"
+	"net"
+	"time"
+
+	logging "github.com/ipfs/go-log/v2"
+	nutil "github.com/textileio/go-threads/net/util"
+	"github.com/textileio/go-threads/util"
+	analytics "github.com/textileio/textile/v2/api/analyticsd/client"
+	analyticspb "github.com/textileio/textile/v2/api/analyticsd/pb"
+	pb "github.com/textileio/textile/v2/api/filrewardsd/pb"
+	"go.mongodb.org/mongo-driver/bson"
+	"go.mongodb.org/mongo-driver/bson/primitive"
+	"go.mongodb.org/mongo-driver/mongo"
+	"go.mongodb.org/mongo-driver/mongo/options"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+	"google.golang.org/protobuf/types/known/timestamppb"
+)
+
+const (
+	rewardsCollectionName = "filrewards"
+	claimsCollectionName  = "filclaims"
+)
+
+var log = logging.Logger("filrewards")
+
+var rewardTypeMeta = map[pb.RewardType]meta{
+	pb.RewardType_REWARD_TYPE_FIRST_KEY_ACCOUNT_CREATED:    {factor: 3},
+	pb.RewardType_REWARD_TYPE_FIRST_KEY_USER_CREATED:       {factor: 1},
+	pb.RewardType_REWARD_TYPE_FIRST_ORG_CREATED:            {factor: 3},
+	pb.RewardType_REWARD_TYPE_INITIAL_BILLING_SETUP:        {factor: 1},
+	pb.RewardType_REWARD_TYPE_FIRST_BUCKET_CREATED:         {factor: 2},
+	pb.RewardType_REWARD_TYPE_FIRST_BUCKET_ARCHIVE_CREATED: {factor: 2},
+	pb.RewardType_REWARD_TYPE_FIRST_MAILBOX_CREATED:        {factor: 1},
+	pb.RewardType_REWARD_TYPE_FIRST_THREAD_DB_CREATED:      {factor: 1},
+}
```

Comment:
TODO: Decide what the base fil reward should be and decide what all the factors should be.

Reply:
Also TODO: Implement more analytics events to trigger more/new fil rewards events.